tower-http-tls/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+extern crate futures;
+extern crate http;
+extern crate rustls;
+extern crate tokio_io;
+extern crate tokio_rustls;
+extern crate tokio_tcp;
+extern crate tower_http_util;
+extern crate tower_service;
+extern crate webpki;
+extern crate webpki_roots;
+
+use futures::{Future, Poll};
+use http::{
+    uri::{Scheme, Uri},
+    Version,
+};
+use rustls::{ClientConfig, Session};
+use std::net::{SocketAddr, ToSocketAddrs};
+use std::sync::Arc;
+use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_rustls::client::TlsStream;
+use tokio_tcp::TcpStream;
+use tower_http_util::HttpConnection;
+use tower_service::Service;
+use webpki::DNSNameRef;
+use webpki_roots::TLS_SERVER_ROOTS;
+
+const ALPN_H2: &str = "h2";
+
+pub struct TlsConnector {
+    config: Arc<ClientConfig>,
+}
+
+pub struct TlsConnection {
+    inner: TlsStream<TcpStream>,
+}
+
+impl TlsConnector {
+    pub fn new(config: ClientConfig) -> Self {
+        let config = Arc::new(config);
```

Comment:
I prefer it.

```rust
impl From<ClientConfig> for TlsConnector {}

impl From<Arc<ClientConfig>> for TlsConnector {}
```

Reply:
Could you explain why this over just take a `ClientConfig`?

Reply:
I have a global `Arc<ClientConfig>` which avoids having to need `Clone` once `ClientConfig` every time a request is create.